benchmarks/test_massdef.py

```diff
@@ -0,0 +1,77 @@
+import os
+import numpy as np
+import pyccl as ccl
+
+cosmo = ccl.Cosmology(Omega_c=0.25, Omega_b=0.05, Omega_g=0, Omega_k=0,
+                      h=0.7, sigma8=0.8, n_s=0.96, Neff=0, m_nu=0.0,
+                      w0=-1, wa=0, T_CMB=2.7255,
+                      transfer_function='eisenstein_hu')
+dirdat = os.path.dirname(__file__) + '/data/'
+Ms, Rs_200m, Rs_500c, Ms_500c = np.loadtxt(dirdat + 'mdef_bm.txt',
+                                           unpack=True)
+dc = np.loadtxt(dirdat + 'conc_bm.txt', unpack=True)
+Ms = dc[0]
+cs_200m_d = dc[1]
+cs_200c_d = dc[2]
+cs_200m_b = dc[3]
+cs_200c_b = dc[4]
+cs_vir_k = dc[5]
+cs_vir_b = dc[6]
+cs_200c_p = dc[7]
+cs_200c_di = dc[8]
+
+hmd_vir = ccl.halos.MassDefVir()
+hmd_vir_b = ccl.halos.MassDefVir('Bhattacharya13')
+hmd_200m = ccl.halos.MassDef200mat()
+hmd_200m_b = ccl.halos.MassDef200mat('Bhattacharya13')
+hmd_200c = ccl.halos.MassDef200crit()
+hmd_200c_b = ccl.halos.MassDef200crit('Bhattacharya13')
+hmd_200c_p = ccl.halos.MassDef200crit('Prada12')
+hmd_200c_di = ccl.halos.MassDef200crit('Diemer15')
+hmd_500c = ccl.halos.MassDef(500, 'critical')
+
+
+def test_mdef_get_radius():
+    Rs_200m_h = hmd_200m.get_radius(cosmo, Ms, 1.)
+    Rs_500c_h = hmd_500c.get_radius(cosmo, Ms, 1.)
+    assert np.all(np.fabs(Rs_200m_h/Rs_200m-1) < 1E-6)
+    assert np.all(np.fabs(Rs_500c_h/Rs_500c-1) < 1E-6)
+
+
+def test_mdef_get_mass():
+    Ms_h = hmd_200m.get_mass(cosmo, Rs_200m, 1.)
+    assert np.all(np.fabs(Ms_h/Ms-1) < 1E-6)
+    Ms_h = hmd_500c.get_mass(cosmo, Rs_500c, 1.)
+    assert np.all(np.fabs(Ms_h/Ms-1) < 1E-6)
+
+
+def test_mdef_concentration():
+    # Duffy 200 matter
+    cs_200m_dh = hmd_200m.get_concentration(cosmo, Ms, 1.)
+    # Bhattacharya 200 matter
+    cs_200m_bh = hmd_200m_b.get_concentration(cosmo, Ms, 1.)
+    # Duffy 200 critical
+    cs_200c_dh = hmd_200c.get_concentration(cosmo, Ms, 1.)
+    # Bhattacharya 200 critical
+    cs_200c_bh = hmd_200c_b.get_concentration(cosmo, Ms, 1.)
+    # Klypin virial
+    cs_vir_kh = hmd_vir.get_concentration(cosmo, Ms, 1.)
+    # Bhattacharya virial
+    cs_vir_bh = hmd_vir_b.get_concentration(cosmo, Ms, 1.)
+    # Prada 200 critical
+    cs_200c_ph = hmd_200c_p.get_concentration(cosmo, Ms, 1.)
+    # Diemer 200 critical
+    cs_200c_dih = hmd_200c_di.get_concentration(cosmo, Ms, 1.)
+    assert np.all(np.fabs(cs_200m_dh/cs_200m_d-1) < 1E-6)
+    assert np.all(np.fabs(cs_200c_dh/cs_200c_d-1) < 1E-6)
+    assert np.all(np.fabs(cs_200m_bh/cs_200m_b-1) < 3E-2)
+    assert np.all(np.fabs(cs_200c_bh/cs_200c_b-1) < 3E-2)
+    assert np.all(np.fabs(cs_200c_ph/cs_200c_p-1) < 3E-2)
+    assert np.all(np.fabs(cs_200c_dih/cs_200c_di-1) < 3E-2)
+    assert np.all(np.fabs(cs_vir_kh/cs_vir_k-1) < 3E-2)
+    assert np.all(np.fabs(cs_vir_bh/cs_vir_b-1) < 3E-2)
```

Comment:
Why are some of these tolerances so high? This is evaluating a function right? We at least need a comment about it.

Reply:
I copy-pasted a lot. Let me check what the actual tolerances should be. In any case, there's always a ~5-10% error on all of these parametrizations due to uncertainties in halo definition, AFAIK.

Reply:
Alright, I've made these as tight as possible. I can't really get mass functions to agree to better than 1%. This is mostly due to the prefactor dlogsigma/dlogM, which is quite sensitive to the power spectrum parametrization. This is only in the high-mass end and at large redshifts. Again, I think 1% in the mass function is not something to worry about, given existing uncertainties (and expected statistical uncertainties for halos of mass 1E15 Msun at z=1).